devlib/instrument/schedstats.py

```diff
@@ -0,0 +1,166 @@
+#
+# See kernel/sched/stat.c and Documentation/scheduler/sched-stats.txt
+#
+# The names used here are based on the identifiers in the scheduler code.
+
+# Some domain fields are repeated for each idle type
+DOMAIN_MEASURES = []
+for idle_type in ['CPU_IDLE', 'CPU_NOT_IDLE', 'CPU_NEWLY_IDLE']:
+    for lb_measure in [
+            'lb_count',
+            'lb_balanced',
+            'lb_failed',
+            'lb_imbalance',
+            'lb_gained',
+            'lb_hot_gained',
+            'lb_nobusyq',
+            'lb_nobusyg']:
+        DOMAIN_MEASURES.append('{}:{}'.format(lb_measure, idle_type))
+
+DOMAIN_MEASURES += [
+    'alb_count',
+    'alb_failed',
+    'alb_pushed',
+    'sbe_count',
+    'sbe_balanced',
+    'sbe_pushed',
+    'sbf_count',
+    'sbf_balanced',
+    'sbf_pushed',
+    'ttwu_wake_remote',
+    'ttwu_move_affine',
+    'ttwu_move_balance'
+]
+
+CPU_MEASURES = [
+    'yld_count',
+    'legacy_always_zero',
+    'schedule_count',
+    'sched_goidle',
+    'ttwu_count',
+    'ttwu_local',
+    'rq_cpu_time',
+    'run_delay',
+    'pcount'
+]
+
+class SchedstatsInstrument(Instrument):
+    """
+    An instrument for parsing Linux's schedstats
+
+    Creates a channel for each entry in the schedstats file and reports the
+    counts directly from the file. Provides useful names for those channels
+    based on identifiers from the scheduler code.
+    """
+    sysctl_path = '/proc/sys/kernel/sched_schedstats'
+    schedstat_path = '/proc/schedstat'
+
+    def setup(self):
+        try:
+            self.old_sysctl_value = self.target.read_int(self.sysctl_path)
+        except TargetError:
+            if not self.target.file_exists(self.sysctl_path):
+                raise TargetError('schedstats not supported by target. '
+                                  'Ensure CONFIG_SCHEDSTATS is enabled.')
+
+        self.target.write_value(self.sysctl_path, 1)
+
+        # Check version matches
+        lines = self.target.read_value(self.schedstat_path).splitlines()
+        match = re.search(r'version ([0-9]+)', lines[0])
+        if not match or match.group(1) != '15':
+            raise TargetError(
+                'Unsupported schedstat version string: "{}"'.format(lines[0]))
+
+        # Take a sample of the schedstat file to figure out which channels to
+        # create.
+        # We'll create a site for each CPU and a site for each sched_domain.
+        for site, measures in self._get_sample().iteritems():
+            if site.startswith('cpu'):
+                measurement_category = 'schedstat_cpu'
+            else:
+                measurement_category = 'schedstat_domain'
+
+            for measurement_name in measures.keys():
+                # TODO: constructing multiple MeasurementTypes with same
+                # params. Does it matter?
+                measurement_type = MeasurementType(
+                    measurement_name, '', measurement_category)
+                self.add_channel(site=site,
```

Comment:
All of this should be done int the '__init__' -- 'setup()' is not guaranteed to be invoked on every execution. (Set up is intended for persistent changes, such as deploying assets to the target.